timemachine/cpp/src/logsumexp.cu

```diff
@@ -0,0 +1,74 @@
+
+#include "gpu_utils.cuh"
+#include "kernels/k_logsumexp.cuh"
+#include "kernels/kernel_utils.cuh"
+#include "logsumexp.hpp"
+#include "math_utils.cuh"
+#include <cub/cub.cuh>
+
+namespace timemachine {
+
+template <typename RealType>
+LogSumExp<RealType>::LogSumExp(const int N) : N_(N), temp_storage_bytes_(0), d_temp_storage_buffer_(nullptr) {
+    void *dummy_temp = nullptr;
+    RealType *dummy_in = nullptr;
+    size_t max_storage_bytes = 0;
+    cub::DeviceReduce::Max(dummy_temp, max_storage_bytes, dummy_in, dummy_in, N_);
+
+    size_t sum_storage_bytes = 0;
+    cub::DeviceReduce::Sum(dummy_temp, sum_storage_bytes, dummy_in, dummy_in, N_);
+
+    // Allocate the larger of the two intermediate values, as we need to run both max and sum
+    temp_storage_bytes_ = max(max_storage_bytes, sum_storage_bytes);
+    d_temp_storage_buffer_.reset(new DeviceBuffer<char>(temp_storage_bytes_));
+};
+
+template <typename RealType> LogSumExp<RealType>::~LogSumExp(){};
+
+template <typename RealType>
+void LogSumExp<RealType>::sum_device(
+    const int N,
+    RealType *d_values,
+    RealType *d_exp_sum_out, // [2] First value is the max value, the second value is the log sum
+    cudaStream_t stream) {
+    if (N > N_) {
+        throw std::runtime_error(
+            "LogSumExp<RealType>::sum_device(): expected N < N_, got N=" + std::to_string(N) +
+            ", N_=" + std::to_string(N_));
+    }
+    // Compute the max value, to subtract for the log computation
+    gpuErrchk(cub::DeviceReduce::Max(d_temp_storage_buffer_->data, temp_storage_bytes_, d_values, d_exp_sum_out, N));
+
+    int tpb = DEFAULT_THREADS_PER_BLOCK;
+    // TBD: Determine if modifying the values in place is appropriate
+    // TBD: Combine Exp and Sum into a single kernel like energy accumulation
+    k_exp_sub_max<<<ceil_divide(N, tpb), tpb, 0, stream>>>(N, d_exp_sum_out, d_values);
+    gpuErrchk(cudaPeekAtLastError());
+
+    gpuErrchk(
+        cub::DeviceReduce::Sum(d_temp_storage_buffer_->data, temp_storage_bytes_, d_values, d_exp_sum_out + 1, N));
+    // Skips the final `log` and addition of the max value call on the device, as it would be a kernel launch
+}
+
+template <typename RealType>
+void LogSumExp<RealType>::sum_host(const int N, const RealType *h_values, RealType *h_out) {
+    DeviceBuffer<RealType> d_vals(N);
+    d_vals.copy_from(h_values);
+
+    DeviceBuffer<RealType> d_out(2);
+
+    cudaStream_t stream = static_cast<cudaStream_t>(0);
+    this->sum_device(N, d_vals.data, d_out.data, stream);
+    gpuErrchk(cudaStreamSynchronize(stream));
+
+    RealType h_intermediate[2];
```

Comment:
I am going back and forth on whether to run a kernel that performs the summation of the log(x) and the max value or the leave it out. Once we start using this in anger, I predict such a kernel would need to die for performance reasons (launch of a kernel time vs computation time is not great) but on the other hand you have this weird logic in the host code. There is a comment about this here (https://github.com/proteneer/timemachine/blob/f207bad143b6abd3d2264c8aca9c937ac2d5dcaa/timemachine/cpp/src/logsumexp.cu#L50)

Will think about some more.

Reply:
Inclined to leave this out as it is a trivial kernel and likely to be done in the kernel that subtracts the logsumexp from the energies to get the log weights.

Reply:
Whatever is simplest!